### PR TITLE
Rameshwar244 bugfix #97

### PR DESCRIPTION
**Pull request by Rameshwar244**
The fix has resolved the Issue #97. The Plugin shows the correct content for text and html format attachments. The attachments formats such as xml, json, pdf are also supported.
Attachments are added with toggle functionality.

Please check the attached reports for better understanding. 
[JSON_And_Reports_New.zip](https://github.com/trivago/cluecumber-report-plugin/files/2510092/JSON_And_Reports_New.zip)

Java Classes
Embedding.Java
ElementPostProcessor.java

Unit Tests
1) EmbeddingTest.java 
- Added getDecodedDataTest() 
- Added getDecodedDataForXMLTest
- Added getFileEndingTest()
2) ElementPostProcessorTest.java
- Modified postDeserializeTest()

Code formatting
     - Modified Cluecumber.css
     - Modified scenario.ftl
Closing issues
https://github.com/trivago/cluecumber-report-plugin/issues/97

### CODE REVIEW

**Comment by Rameshwar244** [comment]
Hi,
I was not aware of Travis build. I missed some imports due to which got compilation errors.
Fixed the code now. Please check

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 1]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
-                                 style="max-width: 100%"/>
-                        </a>
-                    <#else>
-                        ${attachment.data?html}
-                    </#if>
-                </div>
-            </div>
+                <div class="row w-100 p-3 m-0 scenarioAttachment">
+                	<div class="w-100 text-left m-auto border border-dark">             
+                    	<#if attachment.image>
```
Comment (laxersaz) [inline plugin-code/src/main/resources/template/macros/scenario.ftl]: Personally, I am not a fan of making attachments collapsible (plus it looks inconsistent with the existing design). Also, this behaviour was not described in the description of issue #97 which was just about correctly showing the attachments in AfterHooks.

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 2]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
+						    	<a>&nbsp;&nbsp;&nbsp;&nbsp;Attachment ${index} (${attachment.getFileEnding()})</a>
+						  	</div>
+						  	<div id="attachment-${index}" class="collapse collapsable-details">						  
+                        		<a class="grouped_elements" rel="images" href="attachments/${attachment.filename}">
+                            		<img src="attachments/${attachment.filename}" style="max-width: 100%"/>
+                        		</a>
+                        	</div>
+	                    <#elseif attachment.getMimeType() == "text/html">
+	                    	<div data-toggle="collapse" data-target="#attachment-${index}" class="collapsable-control">
+								<a>&nbsp;&nbsp;&nbsp;&nbsp;Attachment ${index} (${attachment.getFileEnding()})</a>
```
Comment (laxersaz) [inline plugin-code/src/main/resources/template/macros/scenario.ftl]: You should be able to get the file name with just `attachment.filename`.

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 3]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
+							</div>
+							<div id="attachment-${index}" class="collapse collapsable-details">
+								<div class="embedding-content">
+							      	<div class="html-content">
+								 		<iframe seamless="true" sandbox="allow-scripts" src="attachments/${attachment.filename}" srcdoc="${attachment.getDecodedData()}"></iframe>
```
Comment (laxersaz) [inline plugin-code/src/main/resources/template/macros/scenario.ftl]: `seamless` is not supported in any browsers. So this should not be added yet.

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 4]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
+							<div id="attachment-${index}" class="collapse collapsable-details">
+								<div class="embedding-content">
+							      	<div class="html-content">
+								 		<iframe seamless="true" sandbox="allow-scripts" src="attachments/${attachment.filename}" srcdoc="${attachment.getDecodedData()}"></iframe> 						  
+							  		</div>
+							  	</div>
+							</div>
+	                    <#elseif attachment.getMimeType() == "text/plain" || attachment.getMimeType() == "text/xml" || attachment.getMimeType() == "application/json" || attachment.getMimeType() == "application/xml">
+	                    	<div data-toggle="collapse" data-target="#attachment-${index}" class="collapsable-control">
+								<a>&nbsp;&nbsp;&nbsp;&nbsp;Attachment ${index} (${attachment.getFileEnding()})</a>
```
Comment (laxersaz) [inline plugin-code/src/main/resources/template/macros/scenario.ftl]: You should be able to get the file ending with just `attachment.fileEnding`.

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 5]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
 
 <#macro attachments step>
     <#if step.embeddings??>
+    	<#assign index=1>
```
Comment (laxersaz) [inline plugin-code/src/main/resources/template/macros/scenario.ftl]: List items have an automatic index: `${x?index}` so an index variable is not needed.

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 6]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
+							    	<span>Download</span>
+							    </a>
+							</div>
+							<div id="attachment-${index}" class="collapse collapsable-details">						  
+								<pre class="embedding-content">This file format is unknown. Click on Download to get the File.</pre>
+							</div>
+	                    </#if>
+                	</div>
+            	</div>
+            <#assign index++>
```
Comment (laxersaz) [inline plugin-code/src/main/resources/template/macros/scenario.ftl]: This is not needed (see comment above).

File: plugin-code/src/test/java/com/trivago/cluecumber/files/FileIOTest.java [hunk 7]
```diff
@@ -37,7 +37,7 @@ public void writeContentToFileStringInvalidTest() throws Exception {
 
     @Test(expected = FileCreationException.class)
     public void writeContentToFileByteArrayInvalidTest() throws Exception {
-        String path = testFolder.getRoot().getPath().substring(1);
+        String path = testFolder.getRoot().getPath().substring(0);
```
Comment (laxersaz) [inline plugin-code/src/test/java/com/trivago/cluecumber/files/FileIOTest.java]: Why did you change this to `substring(0)`? Because this returns the exact same source string.

File: plugin-code/src/test/java/com/trivago/cluecumber/json/pojo/EmbeddingTest.java [hunk 8]
```diff
@@ -55,4 +57,51 @@ public void isPlainTextTest() {
...
+		embedding.setData(encodeString);
+		assertThat(embedding.getDecodedData(),is("This is getDecodeData() Test !!!"));
+    }
+    
+    @Test
+    public void getDecodedDataForXMLTest() {
+		String originalInput = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><note><to>Tove</to><from>Jani</from><heading>Reminder</heading><body>Don't forget me this weekend!</body></note>";
+		String encodeString = new String(Base64.encodeBase64(originalInput.getBytes()));
+		embedding.setData(encodeString);
+        embedding.setMimeType("text/xml");
```
Comment (laxersaz) [inline plugin-code/src/test/java/com/trivago/cluecumber/json/pojo/EmbeddingTest.java]: Please make sure the indentation is consistent with the rest of the code.

File: plugin-code/src/main/java/com/trivago/cluecumber/json/pojo/Embedding.java [hunk 9]
```diff
@@ -52,13 +56,49 @@ public void setFilename(final String filename) {
     public boolean isImage() {
         return
                 mimeType.equalsIgnoreCase("image/png") ||
```
Comment (laxersaz) [inline plugin-code/src/main/java/com/trivago/cluecumber/json/pojo/Embedding.java]: It would make sense to have the mime types as an enum since they are reused in some places throughout your code.

File: plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java [hunk 10]
```diff
@@ -87,56 +87,30 @@ private void processAttachments(final List<Step> steps, List<ResultMatch> afterH
...
-                fileEnding = ".unknown";
-        }
-
-        byte[] dataBytes = Base64.decodeBase64(embedding.getData().getBytes(StandardCharsets.UTF_8));
-
+    private String saveEmbeddingToFileAndGetFilename(final Embedding embedding) {        
         // Clear attachment data to reduce memory
-        embedding.setData("");
-
+        // embedding.setData("");   // Can not set this to empty string. This is now required in getDecodedData() function in Embedding.java
```
Comment (laxersaz) [inline plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java]: Removing this line causes all encoded attachments (including large Base64 encoded image strings) to remain in memory. This should be reverted and handled differently.

In general, instead of commenting out code, it should be removed instead.

File: docs/css/cluecumber.css [hunk 11]
```diff
@@ -49,8 +49,10 @@ body {
 }
 
 pre {
-    margin: 0;
-    white-space: pre-wrap;
+    margin: 10px;
+    word-break: normal !important;
```
[the remark this answers is not in the document]
Comment (Rameshwar244) [reply docs/css/cluecumber.css]: Can we add new rules in the css like '.html-content ' and '.embedding-content'? As existing rules are not showing the attachment content properly. If No, Is there any other workaround for the same?

[the remark this answers is not in the document]
Comment (laxersaz) [reply docs/css/cluecumber.css]: Yes this is possible where it makes sense.

File: plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java (hunk 10 above)
Comment (Rameshwar244) [reply plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java]: Yes, this takes lots of memory. But If we set "embedding.setData("");" after saving the attachments, it works only for image attachments. In order to render other formats attachments, we must have the encoded data for decoding purpose in memory.

Or we may have to use the saved attachments again. I think, For this to work, we have to again somehow encode and decode attachments for rendering. Please suggest your thoughts on this.

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 1 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: I added this behavior to reduce the scrolling overhead.(Currently, it is inconsistent with the existing design, we will have to work on it)
 
As per my opinion, lots of scrolling is already required to see the scenario steps result and one image attachments. In many cases, there could be requirements to attach scenario execution logs, uploaded files, downloaded files, additional reporting parameters and attachments. This will increase scrolling overhead too much.

File: plugin-code/src/test/java/com/trivago/cluecumber/files/FileIOTest.java (hunk 7 above)
Comment (Rameshwar244) [reply plugin-code/src/test/java/com/trivago/cluecumber/files/FileIOTest.java]: No it doesn't. The test was failing due to  "InvalidPathException.class" instead of expected "FileCreationException.class".
Expected Path :- C:\Users\rameshwar.shiral\Downloads\cluecumber-report-plugin-master
Actual Path :- :\Users\rameshwar.shiral\Downloads\cluecumber-report-plugin-master

File: plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java (hunk 10 above)
Comment (laxersaz) [reply plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java]: For the non-image cases, we could either decode the data and write it back to `data` or introduce a new property to store the decoded data.

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 1 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: Agreed. Should I keep titles for attachments like "Attachment 1 (html)" or remove them too?

Comment (laxersaz) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: I think we can remove them.

File: docs/css/cluecumber.css (hunk 11 above)
[the remark this answers is not in the document]
Comment (Rameshwar244) [reply docs/css/cluecumber.css]: Done

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 12]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
-                    <#else>
-                        ${attachment.data?html}
-                    </#if>
-                </div>
-            </div>
+                <div class="row w-100 p-3 m-0 scenarioAttachment">
+                	<div class="w-100 text-left m-auto border border-dark">             
+                    	<#if attachment.image>
+                    		<div data-toggle="collapse" data-target="#attachment-${index}" class="collapsable-control">
+						    	<a>&nbsp;&nbsp;&nbsp;&nbsp;Attachment ${index} (${attachment.getFileEnding()})</a>
```
[the remark this answers is not in the document]
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: removed this code

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 13]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
-                           href="attachments/${attachment.filename}">
-                            <img src="attachments/${attachment.filename}"
-                                 style="max-width: 100%"/>
-                        </a>
-                    <#else>
-                        ${attachment.data?html}
-                    </#if>
-                </div>
-            </div>
+                <div class="row w-100 p-3 m-0 scenarioAttachment">
```
[the remark this answers is not in the document]
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: Done

File: plugin-code/src/main/resources/template/macros/scenario.ftl [hunk 14]
```diff
@@ -105,22 +105,63 @@ limitations under the License.
...
+                    	<#if attachment.image>
+                    		<div data-toggle="collapse" data-target="#attachment-${index}" class="collapsable-control">
+						    	<a>&nbsp;&nbsp;&nbsp;&nbsp;Attachment ${index} (${attachment.getFileEnding()})</a>
+						  	</div>
+						  	<div id="attachment-${index}" class="collapse collapsable-details">						  
+                        		<a class="grouped_elements" rel="images" href="attachments/${attachment.filename}">
+                            		<img src="attachments/${attachment.filename}" style="max-width: 100%"/>
+                        		</a>
+                        	</div>
+	                    <#elseif attachment.getMimeType() == "text/html">
```
[the remark this answers is not in the document]
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: Done

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 1 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: removed attachments collapsible behavior

File: plugin-code/src/test/java/com/trivago/cluecumber/files/FileIOTest.java (hunk 7 above)
Comment (Rameshwar244) [reply plugin-code/src/test/java/com/trivago/cluecumber/files/FileIOTest.java]: I am confused here. I think, substring(index) starts from zero. So "substring(1);" starts from 2nd letter from path.
path with substring(1); returns below path
:\Users\rameshwar.shiral\AppData\Local\Temp\junit1240953755792720228
path with substring(0); returns below path
C:\Users\rameshwar.shiral\AppData\Local\Temp\junit1240953755792720228.
("C" is missing with substring(1);)

Anyway, Currently I have removed the substring(1) and used below code only.

Please correct me if I am missing something in this regard.

File: plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java (hunk 10 above)
Comment (Rameshwar244) [reply plugin-code/src/main/java/com/trivago/cluecumber/json/postprocessors/ElementPostProcessor.java]: new property to store the decoded data.

File: plugin-code/src/test/java/com/trivago/cluecumber/json/pojo/EmbeddingTest.java (hunk 8 above)
Comment (Rameshwar244) [reply plugin-code/src/test/java/com/trivago/cluecumber/json/pojo/EmbeddingTest.java]: Done

File: plugin-code/src/main/java/com/trivago/cluecumber/json/pojo/Embedding.java (hunk 9 above)
Comment (Rameshwar244) [reply plugin-code/src/main/java/com/trivago/cluecumber/json/pojo/Embedding.java]: Added MimeType.java as enum for MimeTypes

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 5 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: Removed this code

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 2 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: Removed this code

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 3 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: removed this code

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 4 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: removed this code

File: plugin-code/src/main/resources/template/macros/scenario.ftl (hunk 6 above)
Comment (Rameshwar244) [reply plugin-code/src/main/resources/template/macros/scenario.ftl]: Done

**Comment by Rameshwar244** [comment]
I have made the required changes.
Please review the same.
Best,
Rameshwar

**Comment by Rameshwar244** [comment]
Thanks Benjamin for merging the code. I will be waiting for next Cluecumber release.
